### ISSUE TITLE
Provide an arm32v7 image

### ISSUE THREAD

**Issue by elrido** [PrivateBin/docker-nginx-fpm-alpine#3]
Given that both the php image as well as the underlying alpine image support the arm 32bit v7 CPU architecture, it should be possible to build such an image without major changes to the existing Dockerfile.

This would allow Raspberry Pi users to easily install PrivateBin as a docker container for use at home.

**Comment by elrido** [comment]
Note to self: Tag "x.y.z" would be a [manifest](https://docs.docker.com/engine/reference/commandline/manifest/) containing i.e. tags "x.y.z-amd64" and "x.y.z-arm32v7".

**Comment by NoxInmortus** [comment]
Hello, same kind of enhancement :  an arm64 image.

**Comment by elrido** [comment]
In theory the existing Dockerfile should work unchanged on these platforms, as alpine does have these packages for all of these flavours. I haven't yet had time to install docker on my Pie to test it - one of these days...

**Comment by NoxInmortus** [comment]
Well I did not tried on armhf but on arm64 it doesnt work.

Docker logs continuously ` exec user process caused "exec format error"`.

**Comment by elrido** [comment]
That is with an image that you built from the Dockerfile on armhf? It should pull the appropriate alpine image or would we have to have a different "FROM" per architecture?

**Comment by NoxInmortus** [comment]
No I only tried the current Dockerfile on a arm64 virtual machine.
But yes I guess you could use `FROM arm64v8/alpine` and try things out.
see https://hub.docker.com/r/arm64v8/alpine/ for more.

Furthermore, If I allow myself to quote Linuxserver.io dockerfiles :
```
Our images support multiple architectures such as x86-64, arm64 and armhf. We utilise the docker manifest for multi-platform awareness. More information is available from docker [here](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#manifest-list) and our announcement [here](https://blog.linuxserver.io/2019/02/21/the-lsio-pipeline-project/).
```
That would be the cleanest thing to do. But I guess that's more work.

**Comment by elrido** [comment]
I had assumed that since the standard alpine:3.10 comes in an [arm64v8 flavour](https://hub.docker.com/layers/alpine/library/alpine/3.10/images/sha256-db7f3dcef3d586f7dd123f107c93d7911515a5991c4b9e51fa2a43e46335a43e), Docker on that CPU type would pick the right version when building a new image? At least that was one of the reasons why I switched to alpine, thinking it would make this type of thing easier...

**Comment by NoxInmortus** [comment]
That's out of my knowledge, but I honestly don't think that is how it works.

Did you tried successfully before (or someone else) on a arm64 instance?

**Comment by elrido** [comment]
No, as mentioned my goal was to get it to work on my pie. I don't have access to a arm64 environment. I'll update this issue once I get around to that.

**Comment by NoxInmortus** [comment]
Well if you need any tester, just poke me out.

You can grab a arm64 instance at https://www.scaleway.com (pricing is by hour of use or something like that).

**Comment by jordan0day** [comment]
Don't want to pile on here, but I also just tried it on armhf, with the same result:

`standard_init_linux.go:211: exec user process caused "exec format error"`

I also have access to a couple different flavors of arm64-based servers and would be happy to test any changes out for you, once they're ready.

**Comment by jordan0day** [comment]
So I did some more digging on this, and I think @elrido's assumption around things just working unchanged because the upstream alpine is multi-arch appears to be correct. The actual issue is here:

https://github.com/PrivateBin/docker-nginx-fpm-alpine/blob/master/Dockerfile#L42

Those references to `s6-overlay-amd64` in the Dockerfiles appear to be the problem. Once I changed those to s6's armhf releases (for my use case) and re-built the image, everything seems to start up happily.

So, in the end it seems you'll still need to go down the multi-arch build route to actually make this work nicely, but for now, anyone needing a quick fix can just edit the `Dockerfile` and rebuild the image.

**Comment by elrido** [comment]
Argh, so it might have worked with the 1.2.1 and 1.3.0 tags, as those used supervisord on python. Anyhow, I'll see that I change the Dockerfile to select the appropriate s6 overlay based on what `uname -m` reports as the architecture, aiming for x86_64, armhf and arm64 support.

**Comment by elrido** [comment]
Tested the above change only on x86_64. According to a [docker blog post](https://engineering.docker.com/2019/04/multi-arch-images/), arm64 reports as aarch64 in `uname -m`, so matches s6 overlay's release naming, while x86_64 and armv7l need a re-mapping. Let me know if this lets you build the image on your respective platforms.

My goal would be to keep using the docker hub automatic builds for multi-arch images. This may be possible using [buildx](https://docs.docker.com/buildx/working-with-buildx/), when they [upgrade their environment](https://github.com/docker/hub-feedback/issues/1874#issuecomment-530419324) to the 19.03 or later release.

**Comment by NoxInmortus** [comment]
No success with 1.3.0 tag.

I can confirm that arm64 `uname -m` returns aarch64.

Building latest dockerfile works fine, and starts fine.
Some errors to mention maybe :
```
docker logs privatebin 
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/nginx/run: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/php-fpm7/run: Operation not permitted
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[04-Oct-2019 09:28:10] NOTICE: fpm is running, pid 181
[04-Oct-2019 09:28:10] NOTICE: ready to handle connections
```

But I reach privatebin with success ! :D

**Comment by elrido** [comment]
Thank you for testing it.

The errors are the same ones that I also get on amd64 arch. The reason is that s6 (like supervisord) is intended to be run as root user. We are using the following workaround to prepare the fifos and do the chown for it that it needs attempts to create / perform on start-up:
https://github.com/PrivateBin/docker-nginx-fpm-alpine/blob/9b8aa78539f565275cd35bf456e413659ff9c42f/Dockerfile#L51-L57
Given that formerly we pulled in the whole python stack just to run supervisord, the now much smaller image should be faster (to pull, update, start up, etc.), more secure (less attack surface) and hopefully worth the confusing errors on startup.

**Comment by NoxInmortus** [comment]
Glad to know.

Well I will stay on my 1.3.1 local build for now and will wait for the arm64 auto-builds with impatience !

Thanks for your work @elrido

**Comment by elrido** [comment]
They still didn't upgrade the build environment on the docker hub to support buildx, but because I wanted to get some other images built for a k3s cluster on some raspberry pis, I figured out how to setup buildx locally, with the help of a [blog article of docker on the topic](https://www.docker.com/blog/getting-started-with-docker-for-arm-on-linux/).

Using this environment, I managed to build and push a number of images for c11b91da26e7d90f0d8b3e960c8e35d489a6cb2f to the 1.3.4 tag (that is the version before the port 80 removal and switch to IPv6 bind). Since this is a manual process, I'll probably only do this for new PrivateBin releases or urgent security updates.

For (primarily my own) future reference, the full sequence to get this working after installing the buildx plugin on Ubuntu 18.04 was:
```shell
docker run --rm --privileged multiarch/qemu-user-static --reset --persistent yes --credential yes
docker buildx create --name builder
docker buildx use builder
docker buildx build --platform linux/arm/v7,linux/arm/v6,linux/arm64,linux/amd64,linux/386,linux/ppc64le -t privatebin/nginx-fpm-alpine:1.3.4 . --push
```
Missing archs: riscv64 doesn't seem supported on my environment or by the qemu-user-static installer and there was no s6-overlay available for s390. But I got the armv7 one that I wanted to test on k3s.
